### PR TITLE
Ports - Elise

### PR DESCRIPTION
**Pull request by ChubbyCub**
# Hotel
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
:------------- | :-------------
What was a design challenge that you encountered on this project? | I had to decide how many classes I really need to answer the user stories request. I had to decide on the relationships between classes and the messages they send to each other. I had to write meaningful tests to challenge my own assumptions in each method. When I started wave 3, due to the block request, I had to find a way for the room to know if it is part of a block booking. This took me a few trial and errors (such as a boolean value blocked: True/False) before I settled with an array of block intervals. Later, I also created another bug for not deleting the blocked interval from the block list for each room instance when that room is booked and moved to the reservation list.
What was a design decision you made that changed over time over the project? | At first, I made all my classes' attributes readable. Later, I remember what the OO Design Principle book mentioned about not exposing the structure of each class. So I had to go back and re-design my classes again to make sure I only expose the minimum amount of information. I let the scheduler class handle most of the info retrieving tasks. I built in more functions in each class to support the scheduler's information retrieval.
What was a concept you gained clarity on, or a learning that you'd like to share? | I learned that OO Design is an extremely difficult process where we should not just create random variables to hold information whenever we feel like we need that piece of information from a given class. This will create a lot of unnecessary dependencies between classes. Instead, I should think of the best way to have a one-way street of information although sometimes there must be two-way of information exchange to solve the problem. 
What is an example of a _nominal test_ that you wrote for this assignment? What makes it a nominal case? | For the reserve_room_in_block method under the Scheduler class, I wrote a test case to see if the method successfully added a new block reservation to the reservation list. This is considered nominal because I just tried to simulate a normal booking process with date ranges, room_ids in the block and the discount rate. None of the parameters passed into the function were unusual.
What is an example of an _edge case test_ that you wrote for this assignment? What makes it an edge case? | For the reserve_room method under the Scheduler class, I wrote a test case for duplicate booking on the same room. This should raise an argument error. This is considered an edge case because the client will almost always want to see which room is available before they decide on which room to book (calling the get_available_rooms method). So if the client book the same room for the exact same duration with the existing reservations, my program should be able to stop them.
How do you feel you did in writing pseudocode first, then writing the tests and then the code? | I wrote my pseudocode in my design docs. This is very effective for me because I can change my design on my design docs as I write my pseudocodes. This is so much better than coding for real and change my complete codes as my design changes.

### CODE REVIEW

File: lib/block.rb [hunk 1]
```diff
@@ -0,0 +1,18 @@
+module Hotel
+  class Block
+    attr_reader :block_id, :room_ids, :reserved_dates, :discounted_rate
+    MAX_NUM_ROOMS = 5.freeze
```
Comment (kaidamasaki) [inline lib/block.rb]: `Integer`s are already immutable so don't need to be frozen.  Good instinct though!

File: lib/block.rb [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+    private_constant :MAX_NUM_ROOMS
```
Comment (kaidamasaki) [inline lib/block.rb:5]: I like that you made your constant private!

File: lib/room.rb [hunk 3]
```diff
@@ -0,0 +1,86 @@
+module Hotel
+  class Room
+    attr_reader :room_id
+
+    RATE = 200.freeze
```
Comment (kaidamasaki) [inline lib/room.rb]: Again, `Integer`s are already mutable.  

Though, almost any other data type in Ruby should be frozen if you assign it to a constant.

File: lib/time_interval.rb [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+    def overlap?(time_interval)
+      return !(time_interval.check_out <= @check_in ||
+               @check_out <= time_interval.check_in)
+    end
+
+    def has_date?(date)
+      return date >= @check_in && date < @check_out
+    end
+
+    def equals?(time_interval)
```
Comment (kaidamasaki) [inline lib/time_interval.rb:25]: You could override `==` here instead of defining your own `equals?` method to clean up your code a tiny bit.

```suggestion
    def ==(time_interval)
```

File: lib/room.rb [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+        if reservation.has_date?(date)
+          return reservation
+        end
+      end
+      return nil
+    end
+
+    def block_dates(reserved_dates)
+      if !is_available?(reserved_dates)
+        raise ArgumentError, "You cannot block out these dates"
```
Comment (kaidamasaki) [inline lib/room.rb]: Like in `reserve` you should use a custom error here.

File: lib/room.rb [hunk 6]
```diff
@@ -0,0 +1,86 @@
...
+
+    def initialize(room_id)
+      @room_id = room_id
+      @reservations = Array.new
+      @block_intervals = Array.new
+    end
+
+    def reserve(reserved_dates)
+      if !is_available?(reserved_dates)
+        raise ArgumentError, "The room you want to reserve is not available."
```
Comment (kaidamasaki) [inline lib/room.rb]: `ArgumentError` should be reserved for invalid arguments of some kind.  The issue here isn't that `reserved_dates` was invalid but that the room was already booked.

You should `raise` a custom error here something like a `RoomBookedError` or `RoomNotAvailableError`.

File: lib/scheduler.rb [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+      @rooms = Hash.new
+      @blocks = Hash.new
+      num_rooms.times do |i|
+        @rooms[i] = Room.new(i)
+      end
+      @next_block_id = 0
+    end
+
+    def get_room_ids
+      return @rooms.keys.sort
```
Comment (kaidamasaki) [inline lib/scheduler.rb:13]: I like that you sorted the keys for the user here. 😃

File: lib/scheduler.rb [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+      @rooms[room_id].reserve(date_range)
+    end
+
+    def get_reservations(date)
+      list = Array.new
+      @rooms.each_key { |id|
+        r = @rooms[id].get_reservation_on_date(date)
+        if !r.nil?
+          list << r
+        end
```
Comment (kaidamasaki) [inline lib/scheduler.rb]: For cases like this you can use `unless` and trailing conditions for slightly cleaner code (though this is already quite clear).

```ruby
list << r unless r.nil?
```

File: lib/scheduler.rb [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+      return @rooms.keys.sort
+    end
+
+    def reserve_room(room_id, date_range)
+      @rooms[room_id].reserve(date_range)
+    end
+
+    def get_reservations(date)
+      list = Array.new
+      @rooms.each_key { |id|
```
Comment (kaidamasaki) [inline lib/scheduler.rb:22]: Not a big deal but you should use `do`/`end` for multi-line blocks and save `{` and `}` for single line blocks.

File: lib/scheduler.rb [hunk 10]
```diff
@@ -0,0 +1,78 @@
+module Hotel
+  class Scheduler
+    def initialize(num_rooms)
+      @rooms = Hash.new
+      @blocks = Hash.new
+      num_rooms.times do |i|
+        @rooms[i] = Room.new(i)
+      end
+      @next_block_id = 0
```
Comment (kaidamasaki) [inline lib/scheduler.rb:9]: Good variable name.  👍

File: lib/room.rb [hunk 11]
```diff
@@ -0,0 +1,86 @@
...
+      @block_intervals.each do |interval|
+        if interval.overlap?(time_interval)
+          return false
+        end
+      end
+      return true
+    end
+
+    def get_reservation_on_date(date)
+      @reservations.each do |reservation|
```
Comment (kaidamasaki) [inline lib/room.rb]: You can clean this up a little using `find`.

```ruby
@reservations.find { |reservation| reservation.has_date?(date) }
```

File: lib/room.rb [hunk 12]
```diff
@@ -0,0 +1,86 @@
...
+      @block_intervals << reserved_dates
+    end
+
+    def has_blocked_dates?(time_interval)
+      @block_intervals.each do |interval|
+        if interval.equals?(time_interval)
+          return true
+        end
+      end
+      return false
```
Comment (kaidamasaki) [inline lib/room.rb]: You can clean this up with `any?`

```ruby
  def has_blocked_dates?(time_interval)
    return @block_intervals.any? { |interval| interval.equals?(time_interval) }
  end
```

File: spec/time_interval_spec.rb [hunk 13]
```diff
@@ -0,0 +1,152 @@
+require "date"
+require "spec_helper.rb"
```
Comment (kaidamasaki) [inline spec/time_interval_spec.rb]: This should be a `require_relative`.

`guard` does something unusual when it loads files so doesn't cause an error but if you use `rake` or run the file directly with `ruby` you will get an error.

File: lib/room.rb [hunk 14]
```diff
@@ -0,0 +1,86 @@
...
+      @reservations = Array.new
+      @block_intervals = Array.new
+    end
+
+    def reserve(reserved_dates)
+      if !is_available?(reserved_dates)
+        raise ArgumentError, "The room you want to reserve is not available."
+      end
+
+      r = Hotel::Reservation.new(reserved_dates, @room_id, RATE)
```
Comment (kaidamasaki) [inline lib/room.rb]: Ruby will be looking for names locally first, prioritizing local names, since they're all in the same module, there's no need to namespace `Hotel::Reservation`. `Reservation` should just work! Same with the rest of the namespacing in the code.

**Comment by kaidamasaki** [comment]
# Hotel
## What We're Looking For



Feature | Feedback
---     | ---
**Baseline** |
Used git regularly | Yes.
Answer comprehension questions | Yes.
**Design** |
Each class is responsible for a single piece of the program | Yes.
Classes are loosely coupled | Yes.
**Wave 1** |
List rooms | Yes.
Reserve a room for a given date range | Yes.
List reservations for a given date | Yes.
Calculate reservation price | Yes.
Invalid date range produces an error | Yes.
Test coverage | Yes.
**Wave 2** |
View available rooms for a given date range | Yes.
Reserving a room that is not available produces an error | Yes.
Test coverage | Yes.
**Wave 3** |
Create a block of rooms | Yes.
Check if a block has rooms | Yes.
Reserve a room from a block | Yes.
Test coverage | Yes.
**Fundamentals** |
Names variables, classes and modules appropriately | Yes.
Understanding of variable scope - local vs instance | Yes.
Can create complex logical structures utilizing variables | Yes.
Appropriately uses methods to break down tasks into smaller simpler tasks | Very Yes.  You made extremely good use of helper methods to clean up your code.
Understands the differences between class and instance methods | N/A (Design didn't require class methods)
Appropriately uses iterators and `Enumerable` methods | Mostly.  You lean a little too heavily on the `each` method, I've noted a few places other enumerable methods would help make your code clearer/cleaner.
Appropriately writes and utilizes classes | Yes.
Appropriately utilizes modules as a namespace | Yes.
**Wrap Up** |
There is a refactors.txt file | Yes.
The file provides a roadmap to future changes | Somewhat.  The file is mostly a list of enhancements (new requirements) and not refactors (code cleanup).
**Additional Feedback** | Great job!  There are a few small issues but overall your code is clear, clean and concise!
